cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/DefaultKafkaQuotasManager.java

```diff
@@ -0,0 +1,210 @@
+
+        return ops;
+    }
+
+    /**
+     * Based on configuration in {@param quotasPlugin}, it configures the default user quota in Kafka.
+     *
+     * @param reconciliation            Reconciliation marker
+     * @param vertx                     Vert.x instance
+     * @param adminClientProvider       Kafka Admin client provider
+     * @param pemTrustSet               Trust set for TLS authentication in PEM format
+     * @param pemAuthIdentity           Identity for TLS client authentication in PEM format
+     * @param quotasPlugin              Configuration of Kafka quotas plugin
+     *
+     * @return  Future that completes when the default user quota configuration is completed
+     */
+    public static Future<Void> configureDefaultUserQuotasInKafka(
+        Reconciliation reconciliation,
+        Vertx vertx,
+        AdminClientProvider adminClientProvider,
+        PemTrustSet pemTrustSet,
+        PemAuthIdentity pemAuthIdentity,
+        QuotasPlugin quotasPlugin
+    ) {
+        Promise<Void> promise = Promise.promise();
+        LOGGER.debugCr(reconciliation, "Attempt to alter default user quotas in Kafka");
+        String bootstrapHostname = KafkaResources.bootstrapServiceName(reconciliation.name()) + "." + reconciliation.namespace() + ".svc:" + KafkaCluster.REPLICATION_PORT;
+
+        try {
+            LOGGER.debugCr(reconciliation, "Creating AdminClient for setting default quota using {}", bootstrapHostname);
+            Admin kafkaAdmin = adminClientProvider.createAdminClient(bootstrapHostname, pemTrustSet, pemAuthIdentity);
+
+            boolean isNotKafkaPlugin = !(quotasPlugin instanceof QuotasPluginKafka);
+            QuotasPluginKafka quotasPluginKafka = isNotKafkaPlugin ? emptyQuotasPluginKafka() : (QuotasPluginKafka) quotasPlugin;
+
+            List<ClientQuotaAlteration.Op> ops = prepareListOfClientQuotaAlterationOps(quotasPluginKafka);
+
+            return shouldAlterDefaultQuotasConfig(reconciliation, vertx, kafkaAdmin, ops, isNotKafkaPlugin)
+                .compose(shouldUpdateQuotas -> {
+                    if (shouldUpdateQuotas) {
+                        ClientQuotaAlteration clientQuotaAlteration = new ClientQuotaAlteration(DEFAULT_USER_ENTITY, ops);
+                        alterQuotas(reconciliation, vertx, kafkaAdmin, clientQuotaAlteration)
+                            .onComplete(result -> {
+                                if (result.succeeded()) {
+                                    LOGGER.infoCr(reconciliation, "Successfully altered default user quotas");
+                                } else {
+                                    LOGGER.warnCr(reconciliation, "Failed to alter default user quotas", result.cause());
+                                }
+                            });
+                    }
+
+                    kafkaAdmin.close();
+                    promise.complete();
```

Comment:
I think you need to redesign this. You need to have this in `else` and should this also be inside the `onComplete` above, or? Otherwise, how do you know you don't close the client and complete the promise before altering the quotas?

Reply:
Hmm good point.
I thought that having it outside the `onComplete` and the `if` will do it properly.
At least it worked all the time during my testing.
But yeah, I guess I will have to do it the way you mentioned.